### DISCUSSION TITLE
Cloud Composer: Job Not Found

### DISCUSSION THREAD

**Discussion by gmyrianthous** [apache/airflow#24277] [Q&A]
I am trying to migrate from Cloud Composer 1 into Cloud Composer 2 (from Airflow 1.10.15 into Airflow 2.2.5) and when attempting to load data from BigQuery into GCS using the `BigQueryToGCSOperator`

```
BigQueryToGCSOperator(
    task_id='my-task',
    source_project_dataset_table=BQ_TABLE_INPUT,
    destination_cloud_storage_uris=f'{GCS_FOLDER_INPUT}*.jsonl',
    export_format='NEWLINE_DELIMITED_JSON',
    compression=None,
    location='europe-west2'
)
```

the following error is being raised:

```
[2022-06-07, 11:17:01 UTC] {taskinstance.py:1776} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/cloud/transfers/bigquery_to_gcs.py", line 141, in execute
    job = hook.get_job(job_id=job_id).to_api_repr()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/common/hooks/base_google.py", line 439, in inner_wrapper
    return func(self, *args, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py", line 1492, in get_job
    job = client.get_job(job_id=job_id, project=project_id, location=location)
  File "/opt/python3.8/lib/python3.8/site-packages/google/cloud/bigquery/client.py", line 2066, in get_job
    resource = self._call_api(
  File "/opt/python3.8/lib/python3.8/site-packages/google/cloud/bigquery/client.py", line 782, in _call_api
    return call()
  File "/opt/python3.8/lib/python3.8/site-packages/google/api_core/retry.py", line 283, in retry_wrapped_func
    return retry_target(
  File "/opt/python3.8/lib/python3.8/site-packages/google/api_core/retry.py", line 190, in retry_target
    return target()
  File "/opt/python3.8/lib/python3.8/site-packages/google/cloud/_http/__init__.py", line 494, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.NotFound: 404 GET https://bigquery.googleapis.com/bigquery/v2/projects/my-project-name/jobs/airflow_1654592634552749_1896245556bd824c71f31c79d28cdfbe?projection=full&prettyPrint=false: Not found: Job my-project-name:airflow_1654592634552749_1896245556bd824c71f31c79d28cdfbe
```

In fact, `composer-2.0.13-airflow-2.2.5` was working fine, till I upgraded to `composer-2.0.14-airflow-2.2.5`

Any clue what may be the issue here and why it does not work on different composer versions?

**Comment by potiuk** [comment]
I suggest you look at the changelog for the provider: https://airflow.apache.org/docs/apache-airflow-providers-google/stable/index.html#changelog - there are many breaking changes in the way how google providers worked since 1.10 so you will likely find some solution there. Also you might want to look at examples in the docs and maybe yo will find a solution there.

BTW.. Airflow has good and comprehensive documentation that's worth checking on your own before reaching out for help. Strongly recommended.

**Reply by gmyrianthous** [reply]
Thanks for your answer. So, in general (and let's just forget for a second that in a different environment the deprecated `BigQueryToCloudStorageOperator` works), when would this error be reported? 

The only relevant thing I found online was about the location, but in my case I explicitly specify it when instantiating the Operator.

**Reply by potiuk** [reply]
No idea, to be honest. Airlfow has more than 3000 operators. You need to look it up.

**Reply by manfredi-giordano** [reply]
I am getting the same error since today, after having migrated from 2.2.3 to 2.2.5. Everything was working until yesterday so I don't think it's related to the 1.10 -> 2.2 upgrade.

**Reply by potiuk** [reply]
Interesting - maybe Google changed something in their service. Did anyone try latest compatible google provider ? 

What is the result there? Does this work?

**Reply by gmyrianthous** [reply]
@manfredi-giordano Can you please let me know what Cloud Composer version you are running? 

`composer-2.0.13-airflow-2.2.5` was working fine, till I upgraded to `composer-2.0.14-airflow-2.2.5`

**Reply by manfredi-giordano** [reply]
@gmyrianthous I am now using `composer-2.0.15-airflow-2.2.5` and it's not working. My previous working version was a bit older, `composer-2.0.3-airflow-2.2.3`.

@potiuk according to [the doc](https://cloud.google.com/composer/docs/concepts/versioning/composer-versions), the Google provider version coming with the latest Composer version is `apache-airflow-providers-google==2022.5.18+composer` which is a bit ambiguous – I haven't been able to map it to any of the versions in https://airflow.apache.org/docs/apache-airflow-providers-google/stable/index.html. I am now going to try forcing the provider version in Composer to be either `6.8.0` or `7.0.0`.

**Reply by gmyrianthous** [reply]
@manfredi-giordano So given that I was having a working Airflow version `v2.2.5` it looks like it's more of a `cloud-composer` version problem. 

+1 on your comments regarding `apache-airflow-providers-google`. Please let me know if any specific version works for you.

**Reply by manfredi-giordano** [reply]
From the [Cloud Composer images](https://cloud.google.com/composer/docs/concepts/versioning/composer-versions#images) list, it would appear that:
- `composer-2.0.13` comes with `apache-airflow-providers-google==6.8.0`
- `composer-2.0.14` onwards come with `apache-airflow-providers-google==2022.5.18+composer`

This would suggest that the provider version `6.8.0` should be safe to use. Unfortunately, I haven't been able to install it as an additional Composer dependency as some of its requirements clash with packages installed by default on the cluster.

**Reply by gmyrianthous** [reply]
@manfredi-giordano Same here 🤔

**Reply by gmyrianthous** [reply]
Checking to see whether [this recent refactoring](https://github.com/apache/airflow/pull/22506) -or [this more recent commit](https://github.com/apache/airflow/pull/23165)-  are related to the issue we are having.

**Reply by manfredi-giordano** [reply]
@gmyrianthous these look spot on. The change from ` hook.insert_job` to `hook.run_extract` in the first pull request seems to be what causes the now present warning
```WARNING - /opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py:1942: DeprecationWarning: This method is deprecated. Please use `BigQueryHook.insert_job` method.``` that wasn't being logged before.

The addition of a `hook.get_job` call after `hook.run_extract` in the second pull request might be what causes the actual ```google.api_core.exceptions.NotFound: 404``` error. This happens even if, in my experience, the BigQuery job does get created (and succeeds) before the library attempts to retrieve it by ID.

**Reply by gmyrianthous** [reply]
Yeap, I also noticed that the files are being transferred anyway, but the Task will fail.

**Reply by potiuk** [reply]
I tihnk this should be fixed in a similar way to: https://github.com/apache/airflow/pull/24165

**Comment by manfredi-giordano** [comment]
As a workaround, adding a custom Composer dependency pointing to a working version of the `google` provider works – running the operator doesn't fail anymore.

I've successfully tested it with `apache-airflow-providers-google==6.8.0`.
Note that because of incompatibilities with packages already installed in Composer clusters, you may need to also define the custom package `proto-plus==1.18.1`.

**Reply by gmyrianthous** [reply]
Makes sense. 

I wanted to avoid the domino effect of package incompatibilities so for time being I just got rid of the operator and wrote my own `PythonOperator` to perform the data transfer using the API. 🤔

**Reply by shuhoy** [reply]
I think the alternative is to use BigQueryInsertJobOperator.
```
airflow.providers.google.cloud.operators.bigquery.BigQueryInsertJobOperator(
  task_id='your_task',
  configuration = {
        'extract': {
            'sourceTable': {
                'projectId': 'your_project',
                'datasetId': 'your_dataset',
                'tableId': 'your_table',
            },
            'compression': None,
            'destinationUris': 'gs://you_bucket/.../file_name.jsonl',
            'destinationFormat': 'NEWLINE_DELIMITED_JSON'
        }
    },
  gcp_conn_id='your_conn_id',
)
```

**Comment by potiuk** [comment]
FYI. I just merged a workaround for now (not a proper fix yet - as the deprecated function remains) in https://github.com/apache/airflow/pull/24330 and it will be available in RC2 of the new google provider I am releasing (8.0.0).

**Reply by manfredi-giordano** [reply]
Thank you!

**Reply by ACowanCW** [reply]
Thanks for the quick fix here. We're experiencing another issue using `8.0.0rc2`

> Traceback (most recent call last):
>   File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/cloud/transfers/bigquery_to_gcs.py", line 128, in execute
>     conf = job["configuration"]["extract"]["sourceTable"]
> TypeError: 'ExtractJob' object is not subscriptable

It looks like the fix returns the full BigQueryJob but is missing the `to_api_repr()` which returns the dict expected in the previous versions.

Potentially adding  `job = job.to_api_repr()` to just prior to [line 128](https://github.com/potiuk/airflow/blob/d2540a9ffbf65c38fae26e5f931f8f3e03d21abf/airflow/providers/google/cloud/transfers/bigquery_to_gcs.py#L118-L128) would fix this.

**Reply by potiuk** [reply]
Ah... yeah.. Looks like :( . I do not plan to cancel/re-release it I hope RC2 wave. Will get released on Monday and I have a promise from the author that the "proper" fix (including deprecation warning removal) will be implemented about that time, so I might want to release 8.0.1 of just google right after.

**Reply by potiuk** [reply]
Thanks for checking!

**Reply by potiuk** [reply]
@wojsamjan FYI -> when the fix is ready I will release 8.0.1.

**Reply by david-loveholidays** [reply]
Looks like this is still broken in the BigQueryToBigQueryOperator code:
```
[2022-06-28, 22:26:28 UTC] {taskinstance.py:1776} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/cloud/transfers/bigquery_to_bigquery.py", line 144, in execute
    job = hook.get_job(job_id=job_id).to_api_repr()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/common/hooks/base_google.py", line 439, in inner_wrapper
    return func(self, *args, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py", line 1494, in get_job
    job = client.get_job(job_id=job_id, project=project_id, location=location)
  File "/opt/python3.8/lib/python3.8/site-packages/google/cloud/bigquery/client.py", line 2066, in get_job
    resource = self._call_api(
  File "/opt/python3.8/lib/python3.8/site-packages/google/cloud/bigquery/client.py", line 782, in _call_api
    return call()
  File "/opt/python3.8/lib/python3.8/site-packages/google/api_core/retry.py", line 283, in retry_wrapped_func
    return retry_target(
  File "/opt/python3.8/lib/python3.8/site-packages/google/api_core/retry.py", line 190, in retry_target
    return target()
  File "/opt/python3.8/lib/python3.8/site-packages/google/cloud/_http/__init__.py", line 494, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.NotFound: 404 GET https://bigquery.googleapis.com/bigquery/v2/projects/...?projection=full&prettyPrint=false: Not found: Job ...:airflow_...```

**Reply by potiuk** [reply]
Did you try with Provider 8.1.0 ?

**Reply by karakanb** [reply]
I am having the same issue with `BigQueryValueCheckOperator` on 8.2.0. 

```
[2022-08-17, 18:02:17 BST] {warnings.py:109} WARNING - /home/airflow/.local/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py:121: DeprecationWarning: This method will be deprecated. Please use `BigQueryHook.get_client` method
  warnings.warn(

[2022-08-17, 18:02:17 BST] {warnings.py:109} WARNING - /home/airflow/.local/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py:2046: DeprecationWarning: This method is deprecated. Please use `BigQueryHook.insert_job` method.
  warnings.warn(

[2022-08-17, 18:02:17 BST] {bigquery.py:1539} INFO - Inserting job some-job-id
[2022-08-17, 18:02:18 BST] {taskinstance.py:1889} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/sql.py", line 205, in execute
    records = self.get_db_hook().get_first(self.sql)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/common/sql/hooks/sql.py", line 202, in get_first
    return cur.fetchone()
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py", line 2711, in fetchone
    return self.next()
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/bigquery.py", line 2727, in next
    self.service.jobs()
  File "/home/airflow/.local/lib/python3.8/site-packages/googleapiclient/_helpers.py", line 134, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/googleapiclient/http.py", line 915, in execute
    raise HttpError(resp, content, uri=self.uri)
googleapiclient.errors.HttpError: <HttpError 404 when requesting https://bigquery.googleapis.com/bigquery/v2/projects/company-project/queries/some-job-id?alt=json returned "Not found: Job company-project:some-job-id". Details: "Not found: Job company-project:some-job-id">
```

Unfortunately I couldn't upgrade to 8.3.0 because it requires `apache-airflow-providers-common-sql>=1.1.0` and v1.1.0 doesn't seem to be available on PyPi yet.

In the meantime I am trying to find what the fix would be as well, but wanted to share it in case someone else has the solution already.

**Reply by potiuk** [reply]
It's available now.

**Reply by karakanb** [reply]
The issue exists on v8.3.0 as well, I am all ears in case anyone has any other ideas to try.

**Reply by karakanb** [reply]
Just upgraded to Airflow v2.4.1 with v8.4.0, the issue persists, `BigQueryValueCheckOperator` doesn't work without the location specified, even though all the other BigQuery operators work properly. Is there another forum that I can report this to help getting it resolved / collaborate to resolve it myself?

**Reply by potiuk** [reply]
You can create PR.

**Reply by potiuk** [reply]
Or open issue. We will mark it as 'good-first-issue" and maybe someone will pick it up.

**Reply by potiuk** [reply]
Just make sure to describe all details.

**Reply by gmyrianthous** [reply]
@karakanb In the meantime, you can follow this workaround: https://stackoverflow.com/a/72548247/7131757
